### DISCUSSION TITLE
Web Service referencing instance of the class

### DISCUSSION THREAD

**Discussion by palmtreefrb** [miguelgrinberg/microdot#167] [General]
Sudo code below. Can I create a ws that can reference my class functions
    
```python
async def temperature_ws(self):
        while True:
            request = self.app.url_map.append(['GET'], '/meas')
            ws = await websocket_upgrade(request)
            temp = self.GetTempFromClassFunction()
            await ws.send(temp)
            await uasyncio.sleep_ms(3000)
```

**Comment by miguelgrinberg** [comment]
This example code is not really designed to be used with Microdot, so it is hard to give you an answer. What is `self` here?

I'm also not sure I understand the term "class functions". If here you mean regular class methods, then these are available without having to have a `self`, so I don't see what the problem is.

**Comment by palmtreefrb** [comment]
My goal is to implement your services without decorators. The decorated functions do not allow access to "self" attributes and methods of my class that microdot is used in. I was looking for a way to use your library without decorated(wrapped) functions.

Example, I cannot do this...

@app.route('/meas')
@with_websocket
async def meas(self, request, ws):

**Reply by miguelgrinberg** [reply]
That is correct. Microdot uses functions, not methods. Even if this worked, I'm not sure what you expect "self" to be here. Where does it come from? If you think about it you'll realize that Microdot cannot invent an object to attach to a web request, that "self" thing has to come from somewhere.

**Reply by palmtreefrb** [reply]
Here is a better example of how I "need" to implement web socket functionality. I guess I will need to look elsewhere...
Sorry, I can't seem to get the formatting to work

class myclass:
    `def __init__(self, value):
        self.value = value`

    `app = Microdot()
    create_task(self.start_webserver())    
     
    
    @await_ready_state
    async def start_webserver(self):
        self.device.dprint("*********************************")
        self.device.dprint("****** starting web server ******")
        self.device.dprint("*********************************")
        self.app.run()
    

     async def update_temps(self):
        do stuff and return temperature

    
    @app.route('/temperature')
    @with_websocket
    async def meas(self, request, ws):
        while True:
            temp = self.update_temps()
            await ws.send(temp)
            await uasyncio.sleep_ms(3000)`

**Reply by miguelgrinberg** [reply]
You sitll have a mysterious "self"  in your code examples that does not come from anywhere, since there are no classes. I'm guessing you are just simplifying the example.

The closest thing I can suggest is to subclass `Microdot`. Something like this:

```python
class MyApp(Microdot):
    async def update_temps(self):
        # do whatever you need to do here

app = MyApp()

@app.route('/temperature')
@with_websocket
async def meas(request, ws):
    while True:
        temp = app.update_temps()
        await ws.send(temp)
        await uasyncio.sleep_ms(3000)

app.run()
```

I wouldn't do this myself, but if you think it is important to have all the web app methods and your own application methods in the same class, this would do it.

**Comment by palmtreefrb** [comment]
Sorry to resurrect this. But after spending time to understand this framework, my question is now. Can you give me an example of how to create a route without a decorator. And a route/webservice without decorator? Is it possible?

**Reply by miguelgrinberg** [reply]
How about this:

```python
from microdot_asyncio import Microdot

app = Microdot()

async def hello(request):
    return 'Hello, world!'

app.route('/')(hello)
app.run(debug=True)
```

**Reply by palmtreefrb** [reply]
doh, of course. Thanks...
![image](https://github.com/miguelgrinberg/microdot/assets/7799495/61c965c1-a58d-4bfb-adaf-5e684fdce5b3)

**Reply by Barriradio** [reply]
```
import uasyncio
from microdot_asyncio import Microdot

# microdot webserver class
class WebServer:
    async def hello1(request):
        return 'Hello world from class'
    
    async def hello2(request):
        return 'Hello world 2 from class'

    def __init__(self):
        self.app = Microdot()
        self.app.route('/1') 	(WebServer.hello1) 
        self.app.route('/2') 	(WebServer.hello2) 
        
    def start_server(self):
        print('Starting microdot server')
        try:
            self.app.run(port=80)
        except:
            self.app.shutdown()
            

ws = WebServer()
ws.start_server()

```
